### DISCUSSION TITLE
Fabric Component Example fails on yarn android (missing type argument for delegate class)

### DISCUSSION THREAD

**Discussion by coolsoftwaretyler** [reactwg/react-native-new-architecture#236] [Q&A] [answered]
Hey folks, 

I've been walking through the [Fabric native components documentation](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/fabric-native-components.md) and hit an error when running the Android side of things (using Kotlin).

As far as I can tell, I have pretty much just copy/pasted the example code line-for-line. The error I get when I run `yarn android` from a consuming app (built with React Native CLI, using RN version 0.75.4, with New Architecture enabled) is:

```sh
e: file:///Users/tylerwilliams/react-native-components-but-worse/example/node_modules/RTNCenteredText/android/src/main/java/com/rtncenteredtext/CenteredTextManager.kt:14:55 2 type arguments expected for class RTNCenteredTextManagerDelegate<T : View!, U : BaseViewManagerInterface<T!>!> where U : RTNCenteredTextManagerInterface<T!>!

FAILURE: Build failed with an exception.
```

Here is my file which looks pretty identical:

```kt
package com.rtncenteredtext

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.RTNCenteredTextManagerInterface
import com.facebook.react.viewmanagers.RTNCenteredTextManagerDelegate

@ReactModule(name = CenteredTextManager.NAME)
class CenteredTextManager(context: ReactApplicationContext) : SimpleViewManager<CenteredText>(), RTNCenteredTextManagerInterface<CenteredText> {
  private val delegate: RTNCenteredTextManagerDelegate<CenteredText> = RTNCenteredTextManagerDelegate(this)

  override fun getDelegate(): ViewManagerDelegate<CenteredText> = delegate

  override fun getName(): String = NAME

  override fun createViewInstance(context: ThemedReactContext): CenteredText = CenteredText(context)

  @ReactProp(name = "text")
  override fun setText(view: CenteredText, text: String?) {
    view.text = text
  }

  companion object {
    const val NAME = "RTNCenteredText"
  }
}
```

When I try to add different type arguments like this:

```kt
  private val delegate: RTNCenteredTextManagerDelegate<CenteredText,  RTNCenteredTextManagerInterface<CenteredText>> = RTNCenteredTextManagerDelegate(this)
```

I end up with errors like this:

```sh
e: file:///Users/tylerwilliams/react-native-components-but-worse/example/node_modules/RTNCenteredText/android/src/main/java/com/rtncenteredtext/CenteredTextManager.kt:15:121 Type mismatch: inferred type is RTNCenteredTextManagerInterface<CenteredText> but BaseViewManagerInterface<TypeVariable(T)!>! was expected
e: file:///Users/tylerwilliams/react-native-components-but-worse/example/node_modules/RTNCenteredText/android/src/main/java/com/rtncenteredtext/CenteredTextManager.kt:15:121 Type mismatch: inferred type is RTNCenteredTextManagerInterface<CenteredText> but BaseViewManagerInterface<CenteredText!>! was expected
```

If I flip over to BaseViewManagerInterface<CenteredText!>, I get the opposite error instead.

Is it likely that I missed a step, or did something change on the codegen side and not get updated in the documentation? I know these docs are early and unstable, so I figured I'd ask around/maybe flag some kind of drift. Thanks in advance!

**Comment by cortinico** [comment, marked as answer]
> I've been walking through the [Fabric native components documentation](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/fabric-native-components.md) and hit an error when running the Android side of things (using Kotlin).

Just FYI, we're updating that guide and we're going to merge it updated on the website between today/tomorrow.

**Reply by coolsoftwaretyler** [reply]
Sick! I can't wait. Really appreciate that you folks have had it available for public consumption.

I'll keep an eye out and try the new docs. Thanks for your work!

**Reply by cortinico** [reply]
Yup I'll update here once they're out

**Reply by coolsoftwaretyler** [reply]
I know y'all have probably had quite a busy week. I poked around on the new docs site and found [the updated guide](https://reactnative.dev/docs/fabric-native-components-introduction), which I'm very excited to read!

Posting here in case anyone else stumbles on the thread.

Congrats on the new architecture release! Stoked for the future.

**Comment by mbpictures** [comment]
I'm the maintainer of the library [react-native-pdf-painter](https://github.com/mbpictures/react-native-pdf-painter/) and I'm trying to implement 0.77 support. However, when I change the definition of the delegate to this (like suggested in the updated guide and docs):
```Kotlin
private val mDelegate: PdfAnnotationViewManagerDelegate<PdfAnnotationView, PdfAnnotationViewManager> = PdfAnnotationViewManagerDelegate(this)
```

But I get these errors:
```
e: file:///E:/Projects/DigitalChoirLeader/packages/mobile/node_modules/react-native-pdf-painter/android/src/main/java/com/pdfannotation/PdfAnnotationViewManager.kt:14:78 Type argument is not within its bounds: should be subtype of 'it(BaseViewManagerInterface & com.facebook.react.viewmanagers.PdfAnnotationViewManagerInterface<com.pdfannotation.PdfAnnotationView!>)!'.
e: file:///E:/Projects/DigitalChoirLeader/packages/mobile/node_modules/react-native-pdf-painter/android/src/main/java/com/pdfannotation/PdfAnnotationViewManager.kt:14:139 Argument type mismatch: actual type is 'com.pdfannotation.PdfAnnotationViewManager', but 'U!' was expected.
```
